File: medium/ReorderLinkedList.cpp

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  void reorderList(ListNode *head) {
    if (!head || !head->next)
      return;
    ListNode *mid = findMiddle(head);
    ListNode *temp = mid->next;
    mid->next = nullptr;
    ListNode *reversedMid = reverseList(temp);
    temp = head;
    ListNode *theAns = temp;
    while (temp and reversedMid) {
      ListNode *tempNext = temp->next;
      ListNode *revNext = reversedMid->next;
      temp->next = reversedMid;
      if (!tempNext)
        break;
      reversedMid->next = tempNext;
      temp = tempNext;
      reversedMid = revNext;
    }
  }

  ListNode *findMiddle(ListNode *head) {
    ListNode *f = head, *s = head;

    while (s and s->next) {
      f = f->next;
      s = s->next->next;
    }
    return f;
  }

  ListNode *reverseList(ListNode *head) {

    ListNode *f = nullptr, *s = head, *t = head->next;
    while (s) {
      t = s->next;
      s->next = f;
      f = s;
      s = t;
    }
    return f;
  }
};

File: py/medium/ReorderLinkedList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        mid = self.findMiddle(head)
        pre_second_half = mid.next
        mid.next = None
        second_half = self.reverse(pre_second_half)
        first_half = head
        while first_half and second_half:
            first_next = first_half.next
            second_next = second_half.next
            first_half.next = second_half
            second_half.next = first_next
            first_half = first_next
            second_half = second_next

    def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, cur = None, head
        while cur:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        return prev

    def findMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow
